packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +424,411 @@ contract AmmPool is IBlockReceiver {
-            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            require(update.tokenWeight == (start ? 0 : ctx.tokens[i].weight), "INVALID_TX_DATA");
             // Now approve this AMM update
             update.validUntil = 0xffffffff;
-            bytes32 txHash = AmmUpdateTransaction.hashTx(exchangeDomainSeparator, update);
+            bytes32 txHash = AmmUpdateTransaction.hashTx(ctx.exchangeDomainSeparator, update);
             exchange.approveTransaction(address(this), txHash);
-            numTransactionsConsumed++;
-            // AMM account balance now available onchain
-            ammBalancesInAccount[i] = update.balance;
-            ammBalances[i] = ammBalancesInAccount[i];
+            ctx.numTransactionsConsumed++;
+            if (start) {
+                // AMM account balance now available onchain
+                ctx.ammBalancesInAccount[i] = update.balance;
+                ctx.ammBalances[i] = ctx.ammBalancesInAccount[i];
+            } else {
+                require(ctx.ammBalances[i] == update.balance, "UNEXPECTED_AMM_BALANCE");
+            }
         }
+    }
 
-        // Process work in the queue
-        for (uint n = queuePos; n < queuePos + auxData.numItems; n++) {
-            QueueItem memory item = queue[n];
-            // TODO: all the necessary pool logic, mint/burn LP tokens,...
-            if (item.workType == WorkType.JOIN) {
-                PoolJoin memory join = item.join;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = join.maxAmountsIn[i];
-                    ammBalances[i] = ammBalances[i].add(amount);
-                    // Unlock the amount locked for this join
-                    locked[item.owner][tokens[i].addr] = locked[item.owner][tokens[i].addr].sub(join.maxAmountsIn[i]);
+    function processJoin(
+        Context  memory ctx,
+        PoolJoin memory join,
+        bytes    memory signature
+        )
+        internal
+    {
+        bytes32 poolTxHash = hashPoolJoin(ctx.DOMAIN_SEPARATOR, join);
+        if (signature.length == 0) {
+            require(queue[queuePos].txHash == poolTxHash, "NOT_APPROVED");
+            delete queue[queuePos];
+            queuePos++;
+        } else {
+            require(poolTxHash.verifySignature(join.owner, signature), "INVALID_SIGNATURE");
+            require(join.fromLayer2, "INVALID_DATA");
+        }
+
+        uint poolTotal = totalSupply();
+        uint ratio = BASE;
+        if (poolTotal > 0) {
+            ratio = (join.poolAmountOut * BASE) / poolTotal;
+        } else {
+            // Important for accuracy
+            require(join.poolAmountOut == INITIAL_SUPPLY, "INITIAL_SUPPLY_UNEXPECTED");
+        }
+
+        // Check if the requirements are fulfilled
+        bool valid = true;
+        uint[] memory amounts = new uint[](ctx.tokens.length);
+        for (uint i = 0; i < ctx.tokens.length; i++) {
```

Comment:
should we use:
```solidity
for (uint i = 0; valid && i < ctx.tokens.length; i++) {
```

Reply:
Possible, though that's strange loop condition for a `for`. :) But it's an inexpensive loop so I would just keep it a standard one.